src/compiler/checker.ts

```diff
@@ -15560,7 +15591,11 @@ namespace ts {
                 return getIndexType(instantiateType((<IndexType>type).type, mapper));
             }
             if (flags & TypeFlags.TemplateLiteral) {
-                return getTemplateLiteralType((<TemplateLiteralType>type).texts, instantiateTypes((<TemplateLiteralType>type).types, mapper));
+                const result = getTemplateLiteralType((<TemplateLiteralType>type).texts, instantiateTypes((<TemplateLiteralType>type).types, mapper));

```

Comment:
The rest of this change exposed an issue where we didn't preserve template literal freshness during instantiation (they are the only instantiable freshable type, so maybe that's not surprising), which is fixed here.